blockchain/v0/reactor.go

```diff
@@ -234,7 +234,7 @@ func (bcR *BlockchainReactor) ReceiveEnvelope(e p2p.Envelope) {
 }
 
 func (bcR *BlockchainReactor) Receive(chID byte, peer p2p.Peer, msgBytes []byte) {
-	var msg *bcproto.Message
+	msg := &bcproto.Message{}
 	err := proto.Unmarshal(msgBytes, msg)
```

Comment:
This is fine, alternatively it should allocate if you pass the address of your pointer instead of the pointer itself (here is an example using JSON, but protobuf should do the same https://go.dev/play/p/hcdjfKv14WC)

The problem is without the address it can't actually modify your pointer value.

Reply:
I see, ok. I think I see the issue with what I was considering. Since the nil pointer points at nothing, you can't just update 'what it points at', since it's not actually pointing to a location in memory yet. I was envisioning that I was copying into the location the pointer points at. There is no location yet, so this can't work. This, however, _would_ work if you allocated in the caller. You _could_ reallocate in the function and then copy into the pointer's location, like so:

https://go.dev/play/p/S9VgfE-gauE

Please let me know if I've mistaken what's happening somehow.

Reply:
That's right.

That's why what you wrote in the PR works fine: When you pass in a pointer to an allocated value, the unmarshaler can modify the contents of that value (through the pointer), and that includes allocating submessages, etc., as needed to decode the message. In that case, the caller's pointer doesn't change (and doesn't need to).

Reply:
Ah, I think the `**msg` approach can't work here because the `Unmarshal` function accepts a `proto.Message`, of which `**msg` is not, even though `*msg` is.

Reply:
Oh yes, you're right, the additional type constraint rules that out.